app/serializers/api/tracks_serializer.rb

```diff
@@ -0,0 +1,98 @@
+class API::TracksSerializer
+  include Rails.application.routes.url_helpers
+  include Mandate
+
+  def initialize(tracks, user = nil)
+    @tracks = tracks
+    @user = user
+  end
+
+  def to_hash
+    {
+      tracks: tracks.map do |track|
+        data_for_track(track).merge(user_data_for_track(track))
+      end
+    }
+  end
+
+  private
+  attr_reader :tracks, :user
+
+  def data_for_track(track)
+    {
+      id: track.id,
+      title: track.title,
+      num_concept_exercises: concept_exercise_counts[track.id].to_i,
+      num_practice_exercises: practice_exercise_counts[track.id].to_i,
+      web_url: routes.track_url(track),
+
+      # TODO: Set all three of these
+      icon_url: "https://assets.exercism.io/tracks/ruby-hex-white.png",
+      is_new: true,
+      tags: ["OOP", "Web Dev"]
+    }
+  end
+
+  def user_data_for_track(track)
+    return {} unless user
+
+    {
+      is_joined: joined_track_ids.include?(track.id),
+      num_completed_concept_exercises: completed_concept_exercise_counts[track.id].to_i,
+      num_completed_practice_exercises: completed_practice_exercise_counts[track.id].to_i
+    }
+  end
+
+  memoize
+  def routes
+    Rails.application.routes.url_helpers
+  end
+
+  memoize
```

Comment:
How does the memoize works, as in: what is the scope in which the memoized value is cached?

Reply:
It effectively does `@#{method_name} ||= the_result_of_the_function`. So it's setting an instance variable (so scoped to the instance of the object - or serializer in this case).

The code is [here](https://github.com/iHiD/mandate/blob/master/lib/mandate/memoize.rb) - I'll happily chat you through it if you'd like to take a walk through some Ruby metaprogramming.